File: src/services/vixl/vixl-tauri/theme-files.ts
import { open, save } from '@tauri-apps/plugin-dialog'
import { call, isTauri } from './helpers'

const THEME_DIALOG_FILTERS = [{ name: 'Vixl Theme (JSON)', extensions: ['json'] }]

const requireTauri = (): void => {
  if (!isTauri()) {
    throw new Error('Vixl desktop APIs are only available in the Tauri app')
  }
}

/**
 * Open a single-file picker for shareable `.vixl-theme.json` files.
 * Returns `null` when the user cancels the dialog (a no-op, not an error).
 */
export const pickThemeFile = async (): Promise<string | null> => {
  requireTauri()

  const selected = await open({
    directory: false,
    multiple: false,
    filters: THEME_DIALOG_FILTERS,
  })

  if (selected === null) {
    return null
  }

  return Array.isArray(selected) ? (selected[0] ?? null) : selected
}

/**
 * Open a save picker for a theme file with a sanitized default filename
 * (including the `.vixl-theme.json` extension). Returns `null` on cancel.
 */
export const pickThemeSavePath = async (defaultName: string): Promise<string | null> => {
  requireTauri()

  return save({
    defaultPath: defaultName,
    filters: THEME_DIALOG_FILTERS,
  })
}

export type ThemeFileContent = {
  content: string
  sizeBytes: number
}

/**
 * Read the raw text of a user-selected theme file. The Rust command enforces
 * a pre-parse size cap before any content reaches the webview; JSON parsing
 * and strict schema validation stay in the import service.
 */
export const readThemeFile = (path: string): Promise<ThemeFileContent> =>
  call('read_theme_file', { path })

/**
 * Write canonical theme JSON to a user-selected path atomically
 * (temp file + rename) so interrupted saves never leave truncated files.
 */
export const writeThemeFile = (path: string, content: string): Promise<void> =>
  call('write_theme_file', { path, content })

File: src-tauri/src/commands/theme_files.rs
use std::fs;
use std::path::{Component, Path, PathBuf};

use serde::Serialize;

use super::config::write_atomic;

/// Upper bound for shareable theme files. Enforced in Rust *before* the
/// content is read or parsed so oversized payloads never reach the webview.
pub const MAX_THEME_FILE_BYTES: u64 = 1024 * 1024;

#[derive(Serialize, Debug)]
pub struct ThemeFileContent {
    pub content: String,
    pub size_bytes: usize,
}

fn resolve_theme_path(path: &str) -> Result<PathBuf, String> {
    let path = Path::new(path);
    if path.as_os_str().is_empty() {
        return Err("Theme file path is empty".to_string());
    }
    for component in path.components() {
        if matches!(component, Component::ParentDir) {
            return Err("Path traversal is not allowed".to_string());
        }
    }
    Ok(path.to_path_buf())
}

/// Read a user-selected `.vixl-theme.json` file as raw text.
///
/// Unlike `read_json_file` (which is restricted to `.vixl` directories and
/// parses eagerly), this command is meant for files picked through the native
/// open dialog and returns unparsed text so the frontend can enforce the
/// strict shareable-theme schema itself.
#[tauri::command]
pub fn read_theme_file(path: String) -> Result<ThemeFileContent, String> {
    let path = resolve_theme_path(&path)?;
    let metadata = fs::metadata(&path).map_err(|e| format!("Failed to read theme file: {e}"))?;
    if !metadata.is_file() {
        return Err("Theme path is not a file".to_string());
    }
    if metadata.len() > MAX_THEME_FILE_BYTES {
        return Err(format!(
            "Theme file is too large ({} bytes, limit is {} bytes)",
            metadata.len(),
            MAX_THEME_FILE_BYTES
        ));
    }
    let content =
        fs::read_to_string(&path).map_err(|e| format!("Failed to read theme file: {e}"))?;
    Ok(ThemeFileContent {
        size_bytes: content.len(),
        content,
    })
}

/// Write a canonical theme JSON payload to a user-selected path.
///
/// The payload is validated as JSON and size-capped, then written atomically
/// (temp file + rename) so an interrupted save never leaves a truncated file.
#[tauri::command]
pub fn write_theme_file(path: String, content: String) -> Result<(), String> {
    let path = resolve_theme_path(&path)?;
    if content.len() as u64 > MAX_THEME_FILE_BYTES {
        return Err(format!(
            "Theme file is too large ({} bytes, limit is {} bytes)",
            content.len(),
            MAX_THEME_FILE_BYTES
        ));
    }
    serde_json::from_str::<serde_json::Value>(&content)
        .map_err(|_| "Theme file content is not valid JSON".to_string())?;
    write_atomic(&path, &content)
}
